### PR TITLE
HDDS-16229. Add RocksDB multiGetSkipCache for batched table reads.

### PR DESCRIPTION
**Pull request by SaketaChalamchala**
## What changes were proposed in this pull request?

RocksDB `MultiGet` support through `RocksDatabase`, `RDBTable`, and `TypedTable` on a skip-cache path. Callers can batch point lookups in one RocksDB call instead of looping `get()` / `getSkipCache()`.
Added `multiGet` with `CodecBuffer` in `TypedTable`.

Missing keys return null at the matching list index; order matches the input key list.

**Call-site guidance**

- Typed reads (Recon, OM): `getTable(name, keyCodec, valueCodec)` → `multiGetSkipCache` → decoded objects.
- Raw persisted bytes (SnapDiff/ `SnapshotDiffValueParser`): use `getTable(name)` → `Table<byte[], byte[]>`; no typed decode.
- Large batches: CodecBuffer path pre-allocates `~N × bufferCapacity` direct memory for value buffers (+ key buffers). Callers should cap batch size for very large N.

**Out of scope (follow-ups):** cache-aware Table.multiGet, batch metrics, call-site adoption.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-16229

## How was this patch tested?

Unit Tests

### CODE REVIEW

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java [hunk 1]
```diff
@@ -228,6 +229,24 @@ public VALUE getSkipCache(KEY key) throws RocksDatabaseException, CodecException
     return getFromTable(key);
   }
 
+  @Override
+  public List<VALUE> multiGetSkipCache(List<KEY> keys)
```
Comment (swamirishi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java:234]: Please implement CodeBuffer based multiGets as well otherwise the GC pressure on memory would be bad

https://github.com/facebook/rocksdb/blob/805ac7c887bbebd7e6124bbc7e8d0803aeed8f07/java/src/main/java/org/rocksdb/RocksDB.java#L2435-L2460

You would need to implement a similar function like 
https://github.com/apache/ozone/blob/9fa7fffb4c6245caaad305655a73501bfb5f6bb6/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java#L318-L355
to ensure the value which didn't fit in the first multiget should be called again with a resized CodecBuffer size.
So the first call should do a multiGet with all CodecBuffer keys and based on the status list returned we need to call multiget again for the ones which didn't fit in the initial fit and collate the list.

**Review by swamirishi — CHANGES_REQUESTED** [review]
Thank you for the patch @SaketaChalamchala this is a good improvement to have.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java [hunk 2]
```diff
@@ -228,6 +230,15 @@ public VALUE getSkipCache(KEY key) throws RocksDatabaseException, CodecException
...
+      throws RocksDatabaseException, CodecException {
+    if (keys.isEmpty()) {
```
Comment (jojochuang) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java]: Suggest to guard against null keys to be consistent with Table.multiGetSkipCache()
```suggestion
    if (keys == null || keys.isEmpty()) {
```

File: hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/utils/db/InMemoryTestTable.java [hunk 3]
```diff
@@ -75,6 +76,18 @@ public VALUE get(KEY key) {
     return map.get(key);
   }
 
+  @Override
+  public List<VALUE> multiGetSkipCache(List<KEY> keys) {
+    if (keys.isEmpty()) {
```
Comment (jojochuang) [inline hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/utils/db/InMemoryTestTable.java]:
```suggestion
    if (keys == null || keys.isEmpty()) {
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java [hunk 4]
```diff
@@ -306,6 +317,114 @@ private VALUE getFromTable(KEY key) throws RocksDatabaseException, CodecExceptio
     }
   }
 
+  List<VALUE> multiGetFromTable(List<KEY> keys)
+      throws RocksDatabaseException, CodecException {
+    if (supportCodecBuffer) {
```
Comment (jojochuang) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TypedTable.java:322]: i would refactor here to make it more readable. Too much things happen in this method.

**Comment by jojochuang** [comment]
@SaketaChalamchala — I took a pass at refactoring `TypedTable.multiGetFromTable()` to address the duplicated status handling between the first pass and the oversize retry loop.

**Branch:** https://github.com/jojochuang/ozone/tree/HDDS-16229-refactor  
**Commit:** https://github.com/jojochuang/ozone/commit/0504b5bc8e4

### What changed
- Split the CodecBuffer vs byte[] paths into `multiGetFromTableWithCodecBuffer()` / `multiGetFromTableWithByteArray()`.
- Extracted shared helpers:
  - `applyMultiGetStatuses()` — first pass (decode or queue retry when `requiredSize > capacity`)
  - `retryOversizedMultiGet()` — second batch for oversize keys
  - `validateMultiGetStatus()`, `decodeFromMultiGetStatus()`, `addWritableValueByteBuffer()`
- Retry keys use `keyByteBuffer.duplicate()` so we don't reuse a buffer whose position may have moved.

Behavior should be unchanged; this is a readability/dedup refactor only.

### Validation
- `TestRDBTable`
- `TestRDBTableStore#multiGetSkipCache`
- `TestRDBTableStore#multiGetSkipCacheWithTypedTable`
- checkstyle on `:hdds-server-framework`

Feel free to cherry-pick or adapt if useful.

**Comment by jojochuang** [comment]
Two performance notes on the CodecBuffer `multiGetFromTable` path:

### 1. Upfront allocation: `n × initial` direct memory

The first pass allocates a direct value buffer for **every** key in the batch before calling RocksDB:

```java
final int initialCapacity = bufferCapacity.get(); // default 4 KB
for (KEY key : keys) {
  ...
  final CodecBuffer valueBuffer = CodecBuffer.allocateDirect(initialCapacity);
  ...
}
```

For a batch of size `n`, that is roughly **`n × initialCapacity`** of direct memory reserved up front (plus `n` direct key buffers), even when most values are small or missing. With large batches this can create noticeable direct-memory pressure and GC churn when buffers are closed in the finally blocks.

The byte[] path avoids this — RocksDB allocates only the actual value sizes. Worth calling out in the PR description if callers may pass large key lists (e.g. OM/SCM metadata scans). A possible follow-up would be lazy/per-key allocation or capping batch size at the call site.

### 2. `bufferCapacity` hint not updated after oversize retry

Single-key reads adapt the shared hint when a value exceeds the current capacity:

```java
// getFromTable() — after resize/retry loop
bufferCapacity.increase(required);
```

`multiGetFromTable` reads `bufferCapacity.get()` for the initial pass but **never calls `bufferCapacity.increase()`** when keys are retried with `requiredSize`. On a table whose values are consistently larger than the default (4 KB), every `multiGetSkipCache` call will pay the two-pass penalty (first pass discovers oversize, second pass re-allocates per oversize key), while single `get()` calls on the same table would have warmed the hint.

Consider updating the hint after the retry pass, e.g. `bufferCapacity.increase(Collections.max(retrySizes))`, so subsequent multiGets on that table start with a better initial capacity. Low priority for correctness, but it would align multi-get behavior with the existing single-get adaptive sizing.

**Comment by SaketaChalamchala** [comment]
> @SaketaChalamchala — I took a pass at refactoring `TypedTable.multiGetFromTable()` to address the duplicated status handling between the first pass and the oversize retry loop.
> 
> **Branch:** https://github.com/jojochuang/ozone/tree/HDDS-16229-refactor **Commit:** [jojochuang@0504b5bc8e4](https://github.com/jojochuang/ozone/commit/0504b5bc8e4)

Thanks for the review and refactor @jojochuang. I took the refactor and added minor updates to update the bufferCapacity hint for the table, the null guards and addressed leaks due to partial value buffer close if there is an exception.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java [hunk 5]
```diff
@@ -122,10 +124,32 @@ public byte[] get(byte[] key) throws RocksDatabaseException {
     return db.get(family, key);
   }
 
+  @Override
+  public List<byte[]> multiGetSkipCache(List<byte[]> keys) throws RocksDatabaseException {
+    if (keys == null || keys.isEmpty()) {
+      return Collections.emptyList();
+    }
+    for (byte[] ignored : keys) {
```
Comment (yandrey321) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java:133]: why not just inc to the keys.size()?

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java [hunk 6]
```diff
@@ -122,10 +124,32 @@ public byte[] get(byte[] key) throws RocksDatabaseException {
...
     return db.get(family, key, outValue);
   }
 
+  List<ByteBufferGetStatus> multiGetSkipCache(List<ByteBuffer> keys, List<ByteBuffer> values)
+      throws RocksDatabaseException {
+    if (keys == null || keys.isEmpty()) {
+      return Collections.emptyList();
+    }
+    for (ByteBuffer ignored : keys) {
+      rdbMetrics.incNumDBKeyGets();
```
Comment (yandrey321) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java:149]: why not just inc to the keys.size()?

Comment (SaketaChalamchala) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java:149]: Need to add the `incNumDBKeyGets(n)` getter to the `RDBMetrics`. Maybe we can address this in a follow up PR for metric improvements.

Comment (yandrey321) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java:149]: please open JIRA for this improvement

Comment (jojochuang) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java:149]: yeah makes sense. Though comparing to RocksDB I/O this is probably negligible impact.

**Review by jojochuang — APPROVED** [review]
Approved. This is clean infrastructure plumbing for HDDS-16229 — RocksDB batch reads through `RocksDatabase` → `RDBTable` → `TypedTable`, with CodecBuffer oversize retry, null/empty guards, refactored helpers, and `bufferCapacity` hint update. No existing production call sites; safe to merge as-is.

Follow-ups already noted elsewhere: JIRA for `incNumDBKeyGets(long)` (yandrey321), call-site adoption, cache-aware multiGet.

## Test coverage — good enough to merge; gaps for follow-up

**What is covered today (sufficient for this PR):**
- `TestRDBTable#testMultiGetSkipCacheDelegatesToDatabase` — mock delegation, missing key → null
- `TestRDBTableStore#multiGetSkipCache` — raw byte[] table, hit/miss mix, null/empty list
- `TestRDBTableStore#multiGetSkipCacheWithTypedTable` — CodecBuffer path, small + oversize retry + missing + empty value, `CodecBuffer.assertNoLeaks()`

**Gaps worth adding in a small follow-up (not merge blockers):**

1. **`bufferCapacity` hint warmup** — no test that a second `multiGetSkipCache` on the same table benefits from `bufferCapacity.increase()` after an oversize retry (would catch regressions where every batch pays two-pass penalty).

2. **All keys missing** — batch where every key is absent; expect all nulls, no retry, no exception.

3. **Explicit order contract** — e.g. `[miss, hit, miss, hit]` to lock index alignment beyond the current mixed cases.

4. **CodecBuffer `RDBTable` path** — no direct test of `multiGetSkipCache(List<ByteBuffer>, List<ByteBuffer>)`; only exercised via `TypedTable`.

5. **Helper unit tests** — `applyMultiGetStatuses` / `retryOversizedMultiGet` / `validateMultiGetStatus` only covered indirectly; mocked `ByteBufferGetStatus` lists would exercise edge cases cheaply.

6. **Non-CodecBuffer typed table** — codecs without `supportCodecBuffer` should use `multiGetFromTableWithByteArray`; untested.

7. **Metrics** — no verification of per-key get counter (pending `incNumDBKeyGets(n)` JIRA anyway).

8. **`DatanodeTable` delegation** — override added, no forward test.

9. **Failure path** — bad `ByteBufferGetStatus` (non-Ok with non-null value) → `IllegalStateException`.

Happy to merge; suggest tracking items 1–3 in HDDS-16229 or a tiny test follow-up when call sites land.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java (hunk 6 above)
Comment (SaketaChalamchala) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBTable.java:149]: Created [HDDS-16380](https://issues.apache.org/jira/browse/HDDS-16380)

**Comment by SaketaChalamchala** [comment]
Fixed the log. Added missing relavent tests and resolved the merge conflict.

**Comment by jojochuang** [comment]
Merged. Thanks @SaketaChalamchala @swamirishi @yandrey321
